### PR TITLE
F# templates

### PR DESCRIPTION
**Pull request by sylvanc**

### CODE REVIEW

**Comment by sylvanc** [comment]
This PR doesn't change the `.sln` file. I'm afraid I don't know what should or shouldn't be added to it :)

**Comment by soninaren** [comment]
please add all F# template files to .sln file

File: Templates/GenericWebHook-FSharp/metadata.json [hunk 1]
```diff
@@ -0,0 +1,11 @@
+{
+    "defaultFunctionName": "GenericWebhookFSharp",
+    "description": "$GenericWebhookFSharp_description",
+    "name": "Generic Webhook",
+    "language": "F#",
+    "enabledInTryMode": true,
+    "category": [
+        "$temp_category_core",
+        "$temp_category_api"
```
Comment (soninaren) [inline Templates/GenericWebHook-FSharp/metadata.json]: should only have $temp_category_experimental

File: Templates/BlobTrigger-FSharp/metadata.json [hunk 2]
```diff
@@ -0,0 +1,15 @@
+{
+    "name": "BlobTrigger",
+    "description": "$BlobTriggerFSharp_description",
```
Comment (soninaren) [inline Templates/BlobTrigger-FSharp/metadata.json]: add $BlobTriggerFSharp_description to the resource file
https://github.com/Azure/azure-webjobs-sdk-templates/blob/dev/Resources/Resources.resx

File: Templates/EventHubTrigger-FSharp/metadata.json [hunk 3]
```diff
@@ -0,0 +1,15 @@
+{
+    "defaultFunctionName": "EventHubTriggerCSharp",
+    "description": "$EventHubTriggerFSharp_description",
```
Comment (soninaren) [inline Templates/EventHubTrigger-FSharp/metadata.json]: EventHubTriggerFSharp_description - Add to resource file

File: Templates/FaceLocator-FSharp/metadata.json [hunk 4]
```diff
@@ -0,0 +1,16 @@
+{
+    "defaultFunctionName": "FaceLocatorCSharp",
+    "description": "$FaceLocatorFSharp_description",
```
Comment (soninaren) [inline Templates/FaceLocator-FSharp/metadata.json]: FaceLocatorFSharp_description - add to resource file

File: Templates/GenericWebHook-FSharp/metadata.json [hunk 5]
```diff
@@ -0,0 +1,11 @@
+{
+    "defaultFunctionName": "GenericWebhookFSharp",
+    "description": "$GenericWebhookFSharp_description",
```
Comment (soninaren) [inline Templates/GenericWebHook-FSharp/metadata.json]: GenericWebhookFSharp_description  - add to resource file

File: Templates/GitHubCommenter-FSharp/metadata.json [hunk 6]
```diff
@@ -0,0 +1,10 @@
+{
+    "defaultFunctionName": "GitHubCommenterFSharp",
+    "description": "$FSharp_description",
```
Comment (soninaren) [inline Templates/GitHubCommenter-FSharp/metadata.json]: should be GitHubCommenterFSharp_description
Also should be add to resource file

File: Templates/HttpGET(CRUD)-FSharp/metadata.json [hunk 7]
```diff
@@ -0,0 +1,16 @@
+{
+    "defaultFunctionName": "HttpGET-CRUD-FSharp",
+    "description": "$HttpGETCRUDFSharp_description",
```
Comment (soninaren) [inline Templates/HttpGET(CRUD)-FSharp/metadata.json]: add to resource file

File: Templates/HttpPOST(CRUD)-FSharp/metadata.json [hunk 8]
```diff
@@ -0,0 +1,16 @@
+{
+    "defaultFunctionName": "HttpPOST-CRUD-FSharp",
+    "description": "$HttpPOSTCRUDFSharp_description",
```
Comment (soninaren) [inline Templates/HttpPOST(CRUD)-FSharp/metadata.json]: add to resource file

File: Templates/HttpPUT(CRUD)-FSharp/metadata.json [hunk 9]
```diff
@@ -0,0 +1,16 @@
+{
+    "defaultFunctionName": "HttpPUT-CRUD-FSharp",
+    "description": "$HttpPUTCRUDFSharp_description",
```
Comment (soninaren) [inline Templates/HttpPUT(CRUD)-FSharp/metadata.json]: Add to resource file

File: Templates/HttpTrigger-FSharp/metadata.json [hunk 10]
```diff
@@ -0,0 +1,14 @@
+{
+    "defaultFunctionName": "HttpTriggerFSharp",
+    "description": "$HttpTriggerFSharp_description",
```
Comment (soninaren) [inline Templates/HttpTrigger-FSharp/metadata.json]: Add to resource file

File: Templates/ImageResizer-FSharp/metadata.json [hunk 11]
```diff
@@ -0,0 +1,12 @@
+{
+    "name": "Image resize",
+    "description": "$ImageResizerFSharp_description",
```
Comment (soninaren) [inline Templates/ImageResizer-FSharp/metadata.json]: Add to resource file

File: Templates/EventHubTrigger-FSharp/metadata.json [hunk 12]
```diff
@@ -0,0 +1,15 @@
+{
+    "defaultFunctionName": "EventHubTriggerCSharp",
```
Comment (soninaren) [inline Templates/EventHubTrigger-FSharp/metadata.json]: Should be EventHubTriggerFSharp

File: Templates/FaceLocator-FSharp/metadata.json [hunk 13]
```diff
@@ -0,0 +1,16 @@
+{
+    "defaultFunctionName": "FaceLocatorCSharp",
```
Comment (soninaren) [inline Templates/FaceLocator-FSharp/metadata.json]: should be FaceLocatorFSharp

File: Templates/SendGrid-FSharp/function.json [hunk 14]
```diff
@@ -0,0 +1,18 @@
+﻿{
+    "bindings": [
+        {
+            "type": "queueTrigger",
+            "name": "order",
+            "direction": "in",
+            "queueName": "samples-orders-csharp"
```
Comment (soninaren) [inline Templates/SendGrid-FSharp/function.json]: nit: samples-orders-fsharp

File: Templates/ManualTrigger-FSharp/metadata.json [hunk 15]
```diff
@@ -0,0 +1,9 @@
+{
+    "defaultFunctionName": "ManualTriggerFSharp",
+    "description": "$ManualTriggerFSharp_description",
```
Comment (soninaren) [inline Templates/ManualTrigger-FSharp/metadata.json]: Add to resource file

File: Templates/SaaSFileTrigger-FSharp/metadata.json [hunk 16]
```diff
@@ -0,0 +1,12 @@
+{
+    "name": "SaaSFileTrigger",
+    "description": "$SaasFileTriggerFSharp_description",
```
Comment (soninaren) [inline Templates/SaaSFileTrigger-FSharp/metadata.json]: Add to resource file

File: Templates/SaaSTable-FSharp/metadata.json [hunk 17]
```diff
@@ -0,0 +1,11 @@
+{
+  "name": "SaaSTable",
+  "description": "$SaasTableFSharp_description",
```
Comment (soninaren) [inline Templates/SaaSTable-FSharp/metadata.json]: Add to resource file

File: Templates/SendGrid-FSharp/metadata.json [hunk 18]
```diff
@@ -0,0 +1,18 @@
+{
+    "defaultFunctionName": "SendGridFSharp",
+    "description": "$SendGridFSharp_description",
```
Comment (soninaren) [inline Templates/SendGrid-FSharp/metadata.json]: Add to resource file

File: Templates/ServiceBusQueueTrigger-FSharp/metadata.json [hunk 19]
```diff
@@ -0,0 +1,16 @@
+{
+    "name": "ServiceBusQueueTrigger",
+    "defaultFunctionName": "ServiceBusQueueTriggerFSharp",
+    "description": "$ServiceBusQueueTriggerFSharp_description",
```
Comment (soninaren) [inline Templates/ServiceBusQueueTrigger-FSharp/metadata.json]: Add to resource file

File: Templates/ServiceBusTopicTrigger-FSharp/metadata.json [hunk 20]
```diff
@@ -0,0 +1,17 @@
+{
+    "name": "ServiceBusTopicTrigger",
+    "defaultFunctionName": "ServiceBusTopicTriggerFSharp",
+    "description": "An F# function that will be run whenever a message is added to the specified Service Bus topic",
```
Comment (soninaren) [inline Templates/ServiceBusTopicTrigger-FSharp/metadata.json]: Use a resource string instead

File: Templates/TimerTrigger-FSharp/metadata.json [hunk 21]
```diff
@@ -0,0 +1,14 @@
+{
+    "defaultFunctionName": "TimerTriggerFSharp",
+    "description": "$TimerTriggerFSharp_description",
```
Comment (soninaren) [inline Templates/TimerTrigger-FSharp/metadata.json]: Add to resource file

File: Templates/TimerTrigger-FSharp/readme.md [hunk 22]
```diff
@@ -0,0 +1,11 @@
+# TimerTrigger - F<span>#</span>
+
+The `TimerTrigger` makes it incredibly easy to have your functions executed on a schedule. This sample demonstrates a simple use case of calling your function every 5 minutes.
```
Comment (soninaren) [inline Templates/TimerTrigger-FSharp/readme.md]: This should be every minute, based on the CRON Expression. Looks like C# and node js template also has this text incorrect

File: Templates/HttpTrigger-FSharp/metadata.json [hunk 23]
```diff
@@ -0,0 +1,14 @@
...
+    "name": "HttpTrigger",
+    "language": "F#",
+    "category": [
+        "$temp_category_core",
+        "$temp_category_api"
```
Comment (soninaren) [inline Templates/HttpTrigger-FSharp/metadata.json]: category should only have $temp_category_experimental

File: Templates/ManualTrigger-FSharp/metadata.json [hunk 24]
```diff
@@ -0,0 +1,9 @@
...
+    "name": "ManualTrigger",
+    "trigger": "ManualTrigger",
+    "language": "F#",
+    "enabledInTryMode": true,
+    "category": [ "$temp_category_core" ]
```
Comment (soninaren) [inline Templates/ManualTrigger-FSharp/metadata.json]: category should only have $temp_category_experimental

File: Templates/SendGrid-FSharp/metadata.json [hunk 25]
```diff
@@ -0,0 +1,18 @@
...
+    "name": "SendGrid",
+    "language": "F#",
+    "category": [
+        "$temp_category_samples",
```
Comment (soninaren) [inline Templates/SendGrid-FSharp/metadata.json]: should only be $temp_category_samples

File: Templates/ServiceBusQueueTrigger-FSharp/metadata.json [hunk 26]
```diff
@@ -0,0 +1,16 @@
...
+    "language": "F#",
+    "category": [
+        "$temp_category_core",
```
Comment (soninaren) [inline Templates/ServiceBusQueueTrigger-FSharp/metadata.json]: should only have $temp_category_experimental

File: Templates/ServiceBusTopicTrigger-FSharp/metadata.json [hunk 27]
```diff
@@ -0,0 +1,17 @@
...
+    "language": "F#",
+    "category": [
+        "$temp_category_core",
```
Comment (soninaren) [inline Templates/ServiceBusTopicTrigger-FSharp/metadata.json]: should only have $temp_category_experimental

File: Templates/TimerTrigger-FSharp/metadata.json [hunk 28]
```diff
@@ -0,0 +1,14 @@
...
+    "name": "TimerTrigger",
+    "language": "F#",
+    "category": [
+        "$temp_category_core",
```
Comment (soninaren) [inline Templates/TimerTrigger-FSharp/metadata.json]: should only have $temp_category_experimental

File: Templates/SaaSTable-FSharp/metadata.json [hunk 29]
```diff
@@ -0,0 +1,11 @@
...
+  "defaultFunctionName": "SaasTableFSharp",
+  "language": "F#",
+  "category": [ "$temp_category_experimental" ],
```
Comment (soninaren) [inline Templates/SaaSTable-FSharp/metadata.json]: missing "enabledInTryMode": false,

**Comment by sylvanc** [comment]
Thanks @soninaren ! These changes should hopefully address all your feedback.

**Comment by soninaren** [comment]
Looks good :shipit:

File: Resources/Resources.resx [hunk 30]
```diff
@@ -837,6 +840,9 @@
   <data name="EventHubTriggerCSharp_description" xml:space="preserve">
     <value>A C# function that will be run whenever an event hub receives a new event</value>
   </data>
+  <data name="EventHubTriggerFSharp_description" xml:space="preserve">
+    <value>A F# function that will be run whenever an event hub receives a new event</value>
```
Comment (mathewc) [inline Resources/Resources.resx]: nit: An.

File: Resources/Resources.resx [hunk 31]
```diff
@@ -876,18 +897,27 @@
...
+    <value>An F# function that adds entities to a Storage Table when it receives an HTTP request</value>
+  </data>
   <data name="HttpPOSTCRUDNodeJS_description" xml:space="preserve">
     <value>A Node.js function that adds entities to a Storage Table when it receives an HTTP request</value>
   </data>
   <data name="HttpPUTCRUDC_description" xml:space="preserve">
     <value>A C# function that updates entitiy in a Storage Table when it receives an HTTP request</value>
   </data>
+  <data name="HttpPUTCRUDFSharp_description" xml:space="preserve">
+    <value>An F# function that updates entitiy in a Storage Table when it receives an HTTP request</value>
```
Comment (mathewc) [inline Resources/Resources.resx]: Entity is spelled wrong. I see the source you copied from is also wrong. Can you fix that one as well?

File: Templates/GitHubWebHook-FSharp/run.fsx [hunk 32]
```diff
@@ -0,0 +1,19 @@
...
+let Run(req: HttpRequestMessage, log: TraceWriter) =
+    async {
+        log.Info(sprintf "F# HTTP trigger function processed a request. RequestUri=%s" (req.RequestUri.ToString()))
+
+        let! data = req.Content.ReadAsAsync<obj>() |> Async.AwaitTask
+        let comment = data?comment?body
+
+        if isNull comment then
+            return req.CreateResponse(HttpStatusCode.BadRequest,
+                "No comment data")
```
Comment (mathewc) [inline Templates/GitHubWebHook-FSharp/run.fsx]: nit: can you move this to previous line?

File: Templates/SaaSFileTrigger-FSharp/run.fsx [hunk 33]
```diff
@@ -0,0 +1,3 @@
+let Run(input: string, name: string, output: byref<string>, log: TraceWriter) =
+    log.Info(sprintf "F# SaaS trigger function processed file: %s" name)
```
Comment (mathewc) [inline Templates/SaaSFileTrigger-FSharp/run.fsx]: You should be able to simplify this function by removing the byref parameter, using $return as the name for the binding in function.json, and simply returning input from this function.

File: Templates/SaaSFileTrigger-FSharp/function.json [hunk 34]
```diff
@@ -0,0 +1,18 @@
...
+    {
+      "type": "apiHubFileTrigger",
+      "name": "input",
+      "direction": "in",
+      "path": "input-cs/{name}",
+      "connection": ""
+    },
+    {
+      "type": "apiHubFile",
+      "name": "output",
```
Comment (mathewc) [inline Templates/SaaSFileTrigger-FSharp/function.json]: Use $return for name - see below

File: Templates/SendGrid-FSharp/run.fsx [hunk 35]
```diff
@@ -0,0 +1,38 @@
...
+// }
+
+[<CLIMutable>]
+type Order = {
+    OrderId: string
+    CustomerName: string
+    CustomerEmail: string
+}
+
+let Run(order: Order, message: byref<Mail>, log: TraceWriter) =
```
Comment (mathewc) [inline Templates/SendGrid-FSharp/run.fsx]: Use a $return binding rather than byref

File: Templates/TimerTrigger-CSharp/function.json [hunk 36]
```diff
@@ -5,7 +5,7 @@
             "name": "myTimer",
             "type": "timerTrigger",
             "direction": "in",
-            "schedule": "0 * * * * *"
+            "schedule": "0 */5 * * * *"
```
Comment (mathewc) [inline Templates/TimerTrigger-CSharp/function.json]: 👍

**Comment by sylvanc** [comment]
Thanks @mathewc ! `$return` binding is very nice :)

**Comment by mathewc** [comment]
:ship:

**Pull request merged**